pkg/ebpf/c/tracee.bpf.c

```diff
@@ -1719,12 +1731,34 @@ int BPF_KPROBE(trace_security_file_ioctl)
         unsigned long fd = PT_REGS_PARM3(ctx);
         struct file *bpf_prog_file = get_struct_file_from_fd(fd);
 
-        send_bpf_attach(&p, bpf_prog_file, perf_event_file);
+        send_bpf_perf_attach(&p, bpf_prog_file, perf_event_file);
     }
 
     return 0;
 }
 
+SEC("kprobe/tracepoint_probe_register_prio_may_exist")
+int BPF_KPROBE(trace_tracepoint_probe_register_prio_may_exist)
```

Comment:
1. This function only exists after kernel v5.10, meaning it will fail in any earlier version (and as well fail tracee I believe)
2. There is another kernel function named [tracepoint_probe_register_prio](https://elixir.bootlin.com/linux/v5.10.178/C/ident/tracepoint_probe_register_prio) - what about it?
3. Both of the above functions call tracepoint_add_func](https://elixir.bootlin.com/linux/v5.10.178/C/ident/tracepoint_add_func) which is also available in earlier kernel version - won't it be better to have a kprobe on it instead?

Reply:
There is also `tracepoint_probe_register` (vanilla), which is the one I can see being called by all `register_trace_##name` functions in DECLARE_TRACE macro. This one was added back in 2015.

By the way, a good point was made. Without an instrumentation test for the bpf_attach (which might be pretty easy to be added), we wouldn't get the error Yaniv mentions, maybe we should add a quick instrumentation test (to make sure logic is tested among all kernel versions and arches).

Reply:
the `tracepoint_probe_register_prio_may_exist` exists in kernels v4.19+ https://elixir.bootlin.com/linux/v4.19.282/source/kernel/tracepoint.c#L364
(there was something weird yesterday with Elixir, it couldn't find the func in versions smaller than v5.10 like Yaniv said, but now it can find it).

if i understand correctly:

tracepoint_add_func is called from two paths - 
bpf + raw_tracepoint
user enabled raw_tracepoint under tracefs

we are only interested in the first path which uses `tracepoint_probe_register_prio_may_exist`.

the other path uses the `tracepoint_probe_register` and `tracepoint_probe_register_prio`.

the bpf + raw_tracepoint path starts in the bpf() syscall.
the only functions we can trace in this path:
bpf() -> bpf_raw_tracepoint_open() -> ... -> tracepoint_probe_register_prio_may_exist()

and only `tracepoint_probe_register_prio_may_exist` holds the information we need (bpf program + tracepoint info) and is not vulnerable to TOCTOU.


there may be another option of combining data from `tracepoint_add_func` with a BPF LSMhook (e.g. `security_bpf_prog`). this will remove the need of using `tracepoint_probe_register_prio_may_exist`, but it still add a kprobe, and also add more comlexity.
WDYT?


about the tests - agreed, we should have an instrumentation test for `bpf_attach` event.
i'll try to add the tests today/tomorrow.

Reply:
```
bpf() -> bpf_raw_tracepoint_open() -> ... -> tracepoint_probe_register_prio_may_exist()
```

If this path exists in previous versions (and it looks it does), then I believe we're good =D.